File: backend/models/pin.py
# Pin model
from sqlalchemy import Column, BigInteger, SmallInteger, ForeignKey, String, DOUBLE_PRECISION, Boolean, DateTime, func
from sqlalchemy.orm import relationship

from database.db import Base



class Pin(Base):
    __tablename__ = 'pins'

    pin_id = Column(BigInteger, primary_key=True, autoincrement=True, index=True)
    cat_id = Column(SmallInteger, ForeignKey("categories.cat_id"), nullable=False)
    sub_cat_id = Column(SmallInteger, ForeignKey("sub_categories.sub_cat_id"), nullable=True)
    user_id = Column(BigInteger, ForeignKey("users.user_id"), nullable=False)
    pin_title = Column(String(100), nullable=False)
    pin_description = Column(String(300), nullable=True)
    pin_picture_path = Column(String(500), nullable=True)
    pin_latitude = Column(DOUBLE_PRECISION, nullable=False)
    pin_longitude = Column(DOUBLE_PRECISION, nullable=False)
    pin_isactive = Column(Boolean, nullable=False, default=True, server_default="true")
    pin_expire_at = Column(DateTime, nullable=False)
    created_at = Column(DateTime, nullable=False, default=func.now(), server_default=func.now())

    # cat_id, user_id, pin_title, pin_latitude, pin_longitude, pin_expire_at
    category = relationship("Category", back_populates="pins")
    sub_category = relationship("SubCategory", back_populates="pins")
    user = relationship("User", back_populates="pins")
    reactions = relationship("PinReaction", back_populates="pin")
    reports = relationship("PinReport", back_populates="pin")

    @property
    def pin_author_name(self) -> str:
        """Return display name if set, otherwise first name only (privacy)"""
        if self.user.user_displayname:
            return self.user.user_displayname
        return self.user.user_fname

    @property
    def pin_color(self) -> str:
        return self.category.category_level.cat_level_color

    @property
    def pin_likes(self) -> int:
        _likes = 0
        for reaction in self.reactions:
            if reaction.reaction_value == 1:
                _likes += 1
        return _likes

    @property
    def pin_dislikes(self) -> int:
        _dislikes = 0
        for reaction in self.reactions:
            if reaction.reaction_value == -1:
                _dislikes += 1
        return _dislikes

    @property
    def user_reaction(self):
        return getattr(self, '_user_reaction', None)

    @user_reaction.setter
    def user_reaction(self, value):
        self._user_reaction = value

File: frontend/lib/models/pin.dart
class Pin {
  final int pinId;
  final int catId;
  final int? subCatId;
  final int userId;
  final String pinTitle;
  final String? pinDescription;
  final String? pinPicturePath;
  final double pinLatitude;
  final double pinLongitude;
  final bool pinIsActive;
  final DateTime pinExpireAt;
  final DateTime createdAt;

  Pin({
    required this.pinId,
    required this.catId,
    this.subCatId,
    required this.userId,
    required this.pinTitle,
    this.pinDescription,
    this.pinPicturePath,
    required this.pinLatitude,
    required this.pinLongitude,
    required this.pinIsActive,
    required this.pinExpireAt,
    required this.createdAt,
  });

  factory Pin.fromJson(Map<String, dynamic> json) {
    return Pin(
      pinId: json['pin_id'] as int,
      catId: json['cat_id'] as int,
      subCatId: json['sub_cat_id'] as int?,
      userId: json['user_id'] as int,
      pinTitle: json['pin_title'] as String,
      pinDescription: json['pin_description'] as String?,
      pinPicturePath: json['pin_picture_path'] as String?,
      pinLatitude: (json['pin_latitude'] as num).toDouble(),
      pinLongitude: (json['pin_longitude'] as num).toDouble(),
      pinIsActive: json['pin_isactive'] as bool,
      pinExpireAt: DateTime.parse(json['pin_expire_at'] as String),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'pin_id': pinId,
      'cat_id': catId,
      'sub_cat_id': subCatId,
      'user_id': userId,
      'pin_title': pinTitle,
      'pin_description': pinDescription,
      'pin_picture_path': pinPicturePath,
      'pin_latitude': pinLatitude,
      'pin_longitude': pinLongitude,
      'pin_isactive': pinIsActive,
      'pin_expire_at': pinExpireAt.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
    };
  }

  bool get isExpired => DateTime.now().isAfter(pinExpireAt);
}

class PinReaction {
  final int reactionId;
  final int pinId;
  final int userId;
  final int reactionValue; // 1 for upvote, -1 for downvote
  final DateTime createdAt;

  PinReaction({
    required this.reactionId,
    required this.pinId,
    required this.userId,
    required this.reactionValue,
    required this.createdAt,
  });

  factory PinReaction.fromJson(Map<String, dynamic> json) {
    return PinReaction(
      reactionId: json['reaction_id'] as int,
      pinId: json['pin_id'] as int,
      userId: json['user_id'] as int,
      reactionValue: json['reaction_value'] as int,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}
